File: modules/monitoring/files/check_mysql_connections.php
<?php

/**
 * Monitor the connection usage of a MySQL server.
 *
 * Usage:
 *   php check_mysql_connections.php --host=HOST --user=USER --password=PASSWORD --max-connections=MAX_CONNECTIONS --critical-threshold=CRITICAL_THRESHOLD --warning-threshold=WARNING_THRESHOLD
 *
 * @author Universal Omega
 */

ini_set( 'display_errors', 'stderr' );

// Parse the command-line options
$options = getopt( '', [
	'host:',
	'user:',
	'password:',
	'current-connections:',
	'max-connections:',
	'critical-threshold:',
	'warning-threshold:',
] );

// Check that the required options have been provided
if (
	!isset( $options['host'] ) ||
	!isset( $options['user'] ) ||
	!isset( $options['password'] ) ||
	!isset( $options['max-connections'] ) ||
	!isset( $options['critical-threshold'] ) ||
	!isset( $options['warning-threshold'] )
) {
	die( 'Usage: php check_mysql_connections.php --host=HOST --user=USER --password=PASSWORD --max-connections=MAX_CONNECTIONS --critical-threshold=CRITICAL_THRESHOLD --warning-threshold=WARNING_THRESHOLD' . "\n" );
}

// Parse the options
$host = $options['host'];
$user = $options['user'];
$pass = $options['password'];
$max_connections = (int)$options['max-connections'];
$critical_threshold = (int)$options['critical-threshold'];
$warning_threshold = (int)$options['warning-threshold'];

// Connect to the MySQL server
$conn = mysqli_init();
$success = mysqli_real_connect( $conn, $host, $user, $pass, null, null, null, false );

if ( !$success ) {
	echo 'Connection failed: ' . mysqli_connect_error();
	exit( 2 );
}

// Retrieve the current number of connections from the SHOW STATUS output
$result = $conn->query('SHOW STATUS WHERE Variable_name = "Threads_connected"');

if ( !$result ) {
	echo 'Query failed: ' . $conn->error;
	exit( 2 );
}

$row = $result->fetch_assoc();
$current_connections = (int)$row['Value'];

$result->free();

// Calculate the connection usage percentage
$connection_usage = $current_connections / $max_connections * 100;

// Display a message and exit status based on the connection usage percentage
if ( $connection_usage >= $critical_threshold ) {
	echo 'Critical connection usage: ' . round($connection_usage, 2) . "%\n";
	echo 'Current connections: ' . $current_connections;
	exit( 2 );
} elseif ( $connection_usage >= $warning_threshold ) {
	echo 'Warning connection usage: ' . round($connection_usage, 2) . "%\n";
	echo 'Current connections: ' . $current_connections;
	exit( 1 );
} else {
	echo 'OK connection usage: ' . round($connection_usage, 2) . "%\n";
	echo 'Current connections: ' . $current_connections;
	exit( 0 );
}

File: modules/monitoring/files/check_mysql_connections.py
#! /usr/bin/python3

"""
This script checks the current number of connections to a MySQL server and compares it to a maximum number of connections.
If the connection usage percentage exceeds a specified critical or warning threshold, a message is displayed and the script exits
with a corresponding exit status.

Created by Universal Omega
"""


import argparse
import mysql.connector

parser = argparse.ArgumentParser(description='MySQL connection usage check')

# Add the arguments
parser.add_argument('--host', required=True, help='MySQL host')
parser.add_argument('--user', required=True, help='MySQL user')
parser.add_argument('--password', required=True, help='MySQL password')
parser.add_argument('--max-connections', type=int, required=True, help='Max connections')
parser.add_argument('--critical-threshold', type=int, required=True, help='Critical threshold')
parser.add_argument('--warning-threshold', type=int, required=True, help='Warning threshold')
parser.add_argument('--ssl-key', help='SSL key file')
parser.add_argument('--ssl-cert', help='SSL certificate file')
parser.add_argument('--ssl-ca', help='SSL CA file')
parser.add_argument('--ssl-verify-server-cert', action='store_true', help='Verify server SSL certificate')

# Parse the command-line arguments
args = parser.parse_args()

# Connect to the MySQL server using SSL
conn = mysql.connector.connect(
    host=args.host,
    user=args.user,
    password=args.password,
    ssl_key=args.ssl_key,
    ssl_cert=args.ssl_cert,
    ssl_ca=args.ssl_ca,
    ssl_verify_server_cert=args.ssl_verify_server_cert
)

# Retrieve the current number of connections from the SHOW STATUS output
cursor = conn.cursor()
cursor.execute('SHOW STATUS WHERE Variable_name = "Threads_connected"')
row = cursor.fetchone()
current_connections = int(row[1])

# Calculate the connection usage percentage
connection_usage = current_connections / args.max_connections * 100

# Display a message and exit status based on the connection usage percentage
if connection_usage >= args.critical_threshold:
    print(f'Critical connection usage: {round(connection_usage, 2)}%')
    exit(2)
elif connection_usage >= args.warning_threshold:
    print(f'Warning connection usage: {round(connection_usage, 2)}%')
    exit(1)
else:
    print(f'OK connection usage: {round(connection_usage, 2)}%')
    exit(0)
